packages/language-server/src/plugins/typescript/DocumentSnapshot.ts

```diff
@@ -1,24 +1,125 @@
 import ts from 'typescript';
-import { getScriptKindFromAttributes } from './utils';
-import { TypescriptDocument } from './TypescriptDocument';
+import { getScriptKindFromAttributes, isSvelteFilePath, getScriptKindFromFileName } from './utils';
+import { Fragment, positionAt, offsetAt, Document, extractTag } from '../../lib/documents';
+import { DocumentMapper, IdentityMapper, ConsumerDocumentMapper } from './DocumentMapper';
+import { Position } from 'vscode-languageserver';
+import { SourceMapConsumer, RawSourceMap } from 'source-map';
+import { pathToUrl } from '../../utils';
+import svelte2tsx from 'svelte2tsx';
 
-export interface DocumentSnapshot extends ts.IScriptSnapshot {
-    version: number;
-    scriptKind: ts.ScriptKind;
+export const INITIAL_VERSION = 0;
+
+export class DocumentSnapshot implements ts.IScriptSnapshot {
+    private fragment?: SnapshotFragment;
+
+    static fromDocument(document: Document) {
+        const { tsxMap, text } = DocumentSnapshot.preprocessIfIsSvelteFile(
+            document.uri,
+            document.getText(),
+        );
+
+        return new DocumentSnapshot(
+            document.version,
+            getScriptKindFromAttributes(extractTag(document.getText(), 'script')?.attributes ?? {}),
+            document.getFilePath() || '',
+            text,
+            tsxMap,
+        );
+    }
+
+    static fromFilePath(filePath: string) {
+        const { text, tsxMap } = DocumentSnapshot.preprocessIfIsSvelteFile(
+            pathToUrl(filePath),
+            ts.sys.readFile(filePath) ?? '',
+        );
+
+        return new DocumentSnapshot(
+            INITIAL_VERSION + 1, // ensure it's greater than initial build
+            getScriptKindFromFileName(filePath),
+            filePath,
+            text,
+            tsxMap,
+        );
+    }
+
+    private static preprocessIfIsSvelteFile(uri: string, text: string) {
+        let tsxMap: RawSourceMap | undefined;
+        if (isSvelteFilePath(uri)) {
+            try {
+                const tsx = svelte2tsx(text);
+                text = tsx.code;
+                tsxMap = tsx.map;
+                if (tsxMap) {
+                    tsxMap.sources = [uri];
+                }
+            } catch (e) {
+                text = '';
```

Comment:
I had this approach and found that it gave too man angry red squiggles when the document is in a state where it won't parse.

https://github.com/halfnelson/svelte-type-checker-vscode/blob/svelte-native-support/server/src/DocumentSnapshot.ts#L50
In my plugin I stash the parse error so that when I get a call to get the diagnostics, I can check for the existence of the error and return the parse error  (instead of treating the file as blank). Is the intention to let the svelte plugin point out the location of an parse errors?

Reply:
Good point. At the moment there is no error at all shown if there are parsing errors. With your strategy this could be pointing to parser errors instead, as you said. I will look into it.

Reply:
Done